pageserver/src/storage_sync/index.rs

```diff
@@ -287,47 +356,106 @@ impl IndexPart {
         disk_consistent_lsn: Lsn,
         metadata_bytes: Vec<u8>,
     ) -> Self {
-        Self {
+        let inner = IndexPartV1 {
             timeline_layers,
             missing_layers,
             disk_consistent_lsn,
             metadata_bytes,
+        };
+        Self::V1 {
+            inner,
+            versioned: true,
+        }
+    }
+
+    pub fn timeline_layers(&self) -> &HashSet<RelativePath> {
+        match self {
+            Self::V1 { inner: v1, .. } => &v1.timeline_layers,
         }
     }
 
+    // FIXME: this should probably be missing_layers()
     pub fn missing_files(&self) -> &HashSet<RelativePath> {
-        &self.missing_layers
+        match self {
+            Self::V1 { inner: v1, .. } => &v1.missing_layers,
+        }
+    }
+
+    pub fn disk_consistent_lsn(&self) -> Lsn {
+        match self {
+            Self::V1 { inner: v1, .. } => v1.disk_consistent_lsn,
+        }
+    }
+
+    pub fn metadata_bytes(&self) -> &[u8] {
+        match self {
+            Self::V1 { inner: v1, .. } => &v1.metadata_bytes,
+        }
+    }
+
+    /// Upgrade the description in-place, if needed.
+    ///
+    /// Returns true when the document was updated and serializing it will produce different
+    /// result than what it was deserialized from.
+    pub fn upgrade(&mut self) -> bool {
```

Comment:
I have some concerns related to this method usage, I think if we've deserialized the type, we've already made all work on presenting the data in some compatible form and can try to autoupgrade it.